pkg/controllers/manageagent/manageagent.go

```diff
@@ -115,23 +121,85 @@ func (h *handler) getAgentBundle(ns string, cluster *fleet.Cluster) (runtime.Obj
 			},
 			Resources: []fleet.BundleResource{
 				{
-					Name:    "agent.yaml",
+					Name:    "base/agent.yaml",
 					Content: string(agentYAML),
 				},
-			},
-			Targets: []fleet.BundleTarget{
 				{
-					ClusterSelector: &metav1.LabelSelector{
-						MatchExpressions: []metav1.LabelSelectorRequirement{
-							{
-								Key:      "fleet.cattle.io/non-managed-agent",
-								Operator: metav1.LabelSelectorOpDoesNotExist,
-							},
+					Name:    "base/kustomization.yaml",
+					Content: string(kustomizeBaseBytes),
+				},
+			},
+		},
+	}
+
+	sort.Slice(clusters, func(i, j int) bool {
+		return clusters[i].Name < clusters[j].Name
+	})
+
+	for _, cluster := range clusters {
+		target := fleet.BundleTarget{
+			ClusterSelector: &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      "fleet.cattle.io/non-managed-agent",
+						Operator: metav1.LabelSelectorOpDoesNotExist,
+					},
+				},
+			},
+			ClusterName: cluster.Name,
+		}
+		target.Kustomize = &fleet.KustomizeOptions{
+			Dir: cluster.Name,
+		}
+
+		// set up kustomize to add env patch per cluster
+		kustomizeFile := map[string][]string{
+			"resources": {"../base"},
+			"patches":   {"patch.yaml"},
+		}
+		kustomizeFileBytes, err := kyaml.Marshal(kustomizeFile)
+		if err != nil {
+			return nil, err
+		}
+
+		containerPatch := map[string]interface{}{
+			"name": agent.DefaultName,
+			"env":  cluster.Spec.AgentEnvVars,
+		}
+		if len(cluster.Spec.AgentEnvVars) == 0 {
+			delete(containerPatch, "env")
+		}
```

Comment:
We delete the patch if there's no env? Is it possible to just create the patch if there is env instead?

Reply:
I tried that, kustomize doesn't like that. If the patch and kustomize option is empty, it will try the manifest as raw yaml and there will be some marshalling error.